community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/DeadState.java

```diff
@@ -0,0 +1,372 @@
+    }
+
+    @Override
+    protected KeyFormat<Key> keyFormat()
+    {
+        return keys;
+    }
+
+    private final KeyFormat<Key> keys;
+    final ActiveState.Factory stateFactory;
+
+    private DeadState( KeyFormat<Key> keys, ActiveState.Factory stateFactory )
+    {
+        this.keys = keys;
+        this.stateFactory = stateFactory;
+    }
+
+    static class Stopped<Key> extends DeadState<Key>
+    {
+        Stopped( KeyFormat<Key> keys, ActiveState.Factory stateFactory )
+        {
+            super( keys, stateFactory );
+        }
+
+        @Override
+        String stateName()
+        {
+            return "stopped";
+        }
+
+        @Override
+        ProgressiveState<Key> initialize( RotationStrategy rotation ) throws IOException
+        {
+            Pair<File, KeyValueStoreFile> opened = rotation.open();
+            if ( opened == null )
+            {
+                return new NeedsCreation<>( keyFormat(), stateFactory, rotation );
+            }
+            return new Prepared<>( stateFactory.open( ReadableState.store( keyFormat(), opened.other() ),
+                                                      opened.first() ) );
+        }
+
+        @Override
+        ProgressiveState<Key> stop() throws IOException
+        {
+            return this;
+        }
+    }
+
+    private static class NeedsCreation<Key> extends DeadState<Key>
+            implements Function<ActiveState<Key>, NeedsCreation<Key>>
+    {
+        private final RotationStrategy rotation;
+
+        private NeedsCreation( KeyFormat<Key> keys, ActiveState.Factory stateFactory, RotationStrategy rotation )
+        {
+            super( keys, stateFactory );
+            this.rotation = rotation;
+        }
+
+        @Override
+        ProgressiveState<Key> stop() throws IOException
+        {
+            return new Stopped<>( keyFormat(), stateFactory );
+        }
+
+        @Override
+        String stateName()
+        {
+            return "needs creation";
+        }
+
+        @Override
+        ActiveState<Key> start( DataInitializer<EntryUpdater<Key>> initializer ) throws IOException
+        {
+            if ( initializer == null )
+            {
+                throw new IllegalStateException( "Store needs to be created, and no initializer is given." );
+            }
+            Pair<File, KeyValueStoreFile> created = initialState( initializer );
+            return stateFactory.open( ReadableState.store( keyFormat(), created.other() ), created.first() );
+        }
+
+        private Pair<File, KeyValueStoreFile> initialState( DataInitializer<EntryUpdater<Key>> initializer )
+                throws IOException
+        {
+            long version = initializer.initialVersion();
+            ActiveState<Key> creation = stateFactory.open( ReadableState.empty( keyFormat(), version ), null );
+            try
+            {
+                try ( EntryUpdater<Key> updater = creation.resetter( new ReentrantLock(), new Runnable()
```

Comment:
It might be nicer to read with the `unsafeUpdater` since we don't need the empty `Runnable`